function/authorization.js

```diff
@@ -30,6 +53,54 @@ class GoogleOAuthAuthorizer extends Authorizer {
         return persistentRequest.post(reqOptions)
     }
 
+    validateAudienceOrEmail(audience, email) {
+        // following two loops use for..in instead of Array.forEach to allow breaking out early
+        // does the audience value start with one of our whitelisted prefixes?
+        for (const i in this.audiencePrefixes) {
+            const prefix = this.audiencePrefixes[i];
+            if (audience.startsWith(prefix.toString())) {
+                return true;
+            }
+        }
+        // does the email value end with one of our whitelisted suffixes?
+        for (const i in this.emailSuffixes) {
+            const suffix = this.emailSuffixes[i];
+            if (email.endsWith(suffix.toString())) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    validateUserInfo(userinfo) {
+        // validate all fields exist
+        userInfoKeys.forEach( (key) => {
+            if (!userinfo.hasOwnProperty(key)) {
+                throw new ResponseError(`OAuth token does not include ${key}`, 403);
+            }
+        })
+        // validate verified_email is Boolean and true
+        const verified = userinfo.verified_email;
+        if (!_.isBoolean(verified)) {
+            throw new ResponseError('OAuth token verified_email must be a Boolean.', 403);
+        } else if (!verified) {
+            throw new ResponseError('OAuth token verified_email must be true.', 403);
+        }
+
+        // validate expires_in is numeric and > 0
+        const expires = _.toNumber(userinfo.expires_in);
+        if (isNaN(expires)) {
+            throw new ResponseError('OAuth token expires_in must be a number.', 403);
+        } else if (expires <= 0) {
+            throw new ResponseError(`OAuth token has expired (expires_in: ${expires})`, 403);
+        }
+        // validate audience/email
+        if (!this.validateAudienceOrEmail(userinfo.audience.toString(), userinfo.email.toString())) {
+            throw new ResponseError(`OAuth token has unacceptable audience (${userinfo.audience}) or email (${userinfo.email})`, 403);
```

Comment:
I would say all of the errors in this function should be 401 errors, not 403. In general 401 means we couldn't verify who you are and 403 means we know who you are and you are not allowed to do this.

Reply:
I would change the wording here: receiving this error would make me think I need both to be acceptable.

Reply:
gotcha on 401 vs 403 and wording, will revamp!